scipy/_lib/_array_api.py

```diff
@@ -163,3 +163,45 @@ def copy(x, *, xp=None):
         xp = array_namespace(x)
 
     return as_xparray(x, copy=True, xp=xp)
+
+
+def assert_equal(actual, desired, err_msg='', xp=None):
+    # TODO compare arrays from different libraries
+    if xp is None:
+        xp = array_namespace(actual)
+    if 'cupy' in xp.__name__:
```

Comment:
There was some discussion about how to do this in https://github.com/scipy/scipy/pull/19023#discussion_r1286902145 and https://github.com/scipy/scipy/pull/19023#discussion_r1295885319. Since it looks like this will merge before gh-19005, do you want to move the `is_numpy` and other helpers into this PR?

Reply:
Yep can do. We probably want to match both the 'raw' and `array_api_compat` versions in that case, since lots of tests don't stick to `(actual, desired)` and have `(desired, actual)` instead, where the desired is raw but the actual will be the compat version. E.g https://github.com/scipy/scipy/blob/6da65af8dac2cf9ea4b94a4dc8e73dd7ffe27225/scipy/fft/tests/test_fftlog.py#L95-L108